keycloak-event-emitter/src/main/java/io/cloudtrust/keycloak/kafkaeventemitter/KafkaEventEmitterProviderFactory.java

```diff
@@ -0,0 +1,149 @@
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.provider.ServerInfoAwareProviderFactory;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class KafkaEventEmitterProviderFactory implements EventListenerProviderFactory, ServerInfoAwareProviderFactory {
+    private static final Logger logger = Logger.getLogger(KafkaEventEmitterProviderFactory.class);
+
+    private static final String PROVIDER_NAME = "Kafka Event Emitter";
+    private static final String PROVIDER_ID = "kafka-event-emitter";
+    private static final String PROVIDER_VERSION = "1.0";
+
+    private static final String SECURITY_PROTOCOL_CONFIG = "security.protocol";
+
+    private static final String BUFFER_CAPACITY_CONFIG_KEY = "bufferCapacity";
+    private static final String CLIENT_ID_CONFIG_KEY = "clientId";
+    private static final String BOOTSTRAP_SERVERS_CONFIG_KEY = "bootstrapServers";
+    private static final String EVENT_TOPIC_CONFIG_KEY = "eventTopic";
+    private static final String ADMIN_EVENT_TOPIC_CONFIG_KEY = "adminEventTopic";
+    private static final String SECURITY_PROTOCOL_KEY = "securityProtocol";
+    private static final String SASL_JAAS_CONFIG_KEY = "saslJaasConfig";
+    private static final String SASL_OAUTHBEARER_TOKEN_ENDPOINT_URL_KEY = "saslOauthbearerTokenEndpointUrl";
+    private static final String SASL_MECHANISM_KEY = "saslMechanism";
+    private static final String SNOWFLAKE_KEYCLOAKID_CONFIG_KEY = "keycloakId";
+    private static final String SNOWFLAKE_DATACENTERID_CONFIG_KEY = "datacenterId";
+
+    private Integer keycloakId;
+    private Integer datacenterId;
+    private String eventTopic;
+    private String adminEventTopic;
+
+    private Properties kafkaProperties;
+    private Producer<String, String> producer;
+    private IdGenerator idGenerator;
+    //Queue only used during STARTING state
+    private LinkedBlockingQueue<ProducerRecord<String, String>> pendingEvents;
+    private KafkaEventEmitterState state;
+    private Lock stateLock;
+
+    @Override
+    public EventListenerProvider create(KeycloakSession session) {
+        if(state.isInitialized()) {
+            state.starting();
+            producer = new KafkaProducer<String, String>(kafkaProperties);
+        }
+        return new KafkaEventEmitterProvider(session, producer, eventTopic, adminEventTopic, idGenerator, pendingEvents, state, stateLock);
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        // Kafka producer configuration
+        eventTopic = getStringConfig(config, EVENT_TOPIC_CONFIG_KEY);
+        adminEventTopic = getStringConfig(config, ADMIN_EVENT_TOPIC_CONFIG_KEY);
+
+        kafkaProperties = new Properties();
+        kafkaProperties.put(SECURITY_PROTOCOL_CONFIG, getStringConfig(config, SECURITY_PROTOCOL_KEY));
```

Comment:
Don't forget to rebase your work on master to have latest changes and update keycloak.properties and install.sh with your configuration properties.